### PR TITLE
4964430: (spec) missing IllegalStateException exception requirement for javax.crypto.Cipher.doFinal

### PR DESCRIPTION
**Pull request by blperez01**
Updated IllegalStateException exception requirements for `update`, `doFinal`, `wrap`, and `unwrap`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-4964430](https://bugs.openjdk.org/browse/JDK-4964430): (spec) missing IllegalStateException exception requirement for javax.crypto.Cipher.doFinal (**Bug** - P4)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15727/head:pull/15727` \
`$ git checkout pull/15727`

Update a local copy of the PR: \
`$ git checkout pull/15727` \
`$ git pull https://git.openjdk.org/jdk.git pull/15727/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15727`

View PR using the GUI difftool: \
`$ git pr show -t 15727`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15727.diff">https://git.openjdk.org/jdk/pull/15727.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15727#issuecomment-1718100625)

### CODE REVIEW

**Comment by valeriepeng** [comment]
1) Cipher.updateAAD(...) methods also calls checkCipherState(), so its javadoc should be updated as well?
2) Given the current checkCipherState() impl does not check for states for NullCipher objects, do we want to also update the javax.crypto.NullCipher class javadoc to mention that IllegalStateException is not thrown by various methods?
3) Update copyright year to 2023?

**Comment by blperez01** [comment]
> * Cipher.updateAAD(...) methods also calls checkCipherState(), so its javadoc should be updated as well?

The current documentation for Cipher.updateAAD(...) is:

"@throws IllegalStateException if this {@code Cipher} object is in a wrong state (e.g., has not been initialized), does not accept AAD, or if operating in either GCM or CCM mode and one of the {@code update} methods has already been called for the active encryption/decryption operation"

This seems fairly explicit, but I could change the wording to 

"@throws IllegalStateException if this {@code Cipher} object is in a wrong state (e.g., has not been initialized, or is not in ENCRYPT_MODE or DECRYPT_MODE)"

**Comment by valeriepeng** [comment]
> > * Cipher.updateAAD(...) methods also calls checkCipherState(), so its javadoc should be updated as well?
> 
> The current documentation for Cipher.updateAAD(...) is:
> 
> "@throws IllegalStateException if this {@code Cipher} object is in a wrong state (e.g., has not been initialized), does not accept AAD, or if operating in either GCM or CCM mode and one of the {@code update} methods has already been called for the active encryption/decryption operation"
> 
> This seems fairly explicit, but I could change the wording to
> 
> "@throws IllegalStateException if this {@code Cipher} object is in a wrong state (e.g., has not been initialized, or is not in ENCRYPT_MODE or DECRYPT_MODE)"

We still need the parts about not accepting AAD and for GCM or CCM mode...

**Comment by blperez01** [comment]
> * Given the current checkCipherState() impl does not check for states for NullCipher objects, do we want to also update the javax.crypto.NullCipher class javadoc to mention that IllegalStateException is not thrown by various methods?

Current documentation is:

"The {@code NullCipher} class is a class that provides an "identity cipher" -- one that does not transform the plain text.  As a consequence, the ciphertext is identical to the plaintext.  All initialization methods do nothing, while the blocksize is set to 1 byte."

Would it be sufficient for me to add "Unlike other ciphers, the {@code NullCipher} will not throw an {@code IllegalStateException} when in an incorrect state."

File: src/java.base/share/classes/javax/crypto/Cipher.java [hunk 1]
```diff
@@ -1851,7 +1851,8 @@ private void checkCipherState() {
      * new block
      *
      * @throws IllegalStateException if this {@code Cipher} object is in a
-     * wrong state (e.g., has not been initialized)
+     * wrong state (e.g., has not been initialized, or is not
+     * in ENCRYPT_MODE or DECRYPT_MODE)
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/Cipher.java:1855]: Use code format when referencing the modes, ex: change this line to:

`* in {@code ENCRYPT_MODE} or {@code DECRYPT_MODE})`

Same comment applies to all other applicable lines below.

File: src/java.base/share/classes/javax/crypto/NullCipher.java [hunk 2]
```diff
@@ -30,7 +30,9 @@
  * "identity cipher" -- one that does not transform the plain text.  As
  * a consequence, the ciphertext is identical to the plaintext.  All
  * initialization methods do nothing, while the blocksize is set to 1
- * byte.
+ * byte.  Unlike other ciphers, the {@code NullCipher} will not throw an
+ * {@code IllegalStateException} when calling {@code Cipher} methods
+ * in an incorrect state.
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/NullCipher.java:35]: Suggest rewording last part as "... when {@code Cipher} methods are called in an incorrect state." I also think we should say a NullCipher has no state. So how about:

"Unlike other ciphers, the {@code NullCipher} has no state, and will not throw an {@code IllegalStateException} when {@code Cipher} methods are called in an incorrect state."

**Review by seanjmullan — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/javax/crypto/NullCipher.java (hunk 2 above)
Comment (ferakocz) [reply src/java.base/share/classes/javax/crypto/NullCipher.java:35]: Well, if it has no state it cannot be in an incorrect state so the " in an incorrect state" part can be omitted.

Comment (seanjmullan) [reply src/java.base/share/classes/javax/crypto/NullCipher.java:35]: Good point!

**Review by seanjmullan — APPROVED** [review]

**Review by valeriepeng — APPROVED** [review]
